### PR TITLE
chore: navigation overhaul cherry picks - WPB-6647

### PR DESCRIPTION
**Pull request by caldrian**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-6647" title="WPB-6647" target="_blank">TopicWPB-6647</a>  [iOS] Navigational Overhaul
  </td></table> 

### Issue

This PR copies the changes of WireUI from the navigation overhaul epic branch to develop.
The connect/start conversation UI presentation has been changed to modal.

### Testing

The application shouldn't have changed significantly yet.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [x] Make sure you use the API for UI elements that support large fonts.
- [x] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [x] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

File: WireUI/Tests/WireSidebarUITests/Resources/ReferenceImages/SidebarMenuItemSnapshotTests/testColorSchemeVariants.light.png
Comment (caldrian) [inline WireUI/Tests/WireSidebarUITests/Resources/ReferenceImages/SidebarMenuItemSnapshotTests/testColorSchemeVariants.light.png:1]: Since blue is the default accent color I set the default selection color to gray.

File: WireUI/Tests/WireMainNavigationUITests/Resources/ReferenceImages/MainSplitViewControllerTests/testSidebarAppearancePortrait.light.png
Comment (caldrian) [inline WireUI/Tests/WireMainNavigationUITests/Resources/ReferenceImages/MainSplitViewControllerTests/testSidebarAppearancePortrait.light.png:1]: In portrait mode the sidebar overlaps the conversation list.

File: WireFoundation/Sources/WireTesting/SDKExtensions/UIKit/UISplitViewControllerDisplayMode+CustomDebugStringConvertible.swift [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            "secondaryOnly"
+        case .twoBesideSecondary:
+            "twoBesideSecondary"
+        case .twoDisplaceSecondary:
+            "twoDisplaceSecondary"
+        case .twoOverSecondary:
+            "twoOverSecondary"
+        @unknown default:
+            "unknown"
+        }
```
Comment (netbe) [inline WireFoundation/Sources/WireTesting/SDKExtensions/UIKit/UISplitViewControllerDisplayMode+CustomDebugStringConvertible.swift]: would String(describing: self) works?

File: WireUI/Sources/WireMainNavigationUI/Containers/DoubleColumnContainerViewController.swift [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            view.bottomAnchor.constraint(equalTo: primaryNavigationController.view.bottomAnchor),
+            primaryNavigationController.view.widthAnchor.constraint(equalToConstant: primaryColumnWidth),
+
+            secondaryNavigationController.view.leadingAnchor.constraint(equalTo: borderView.trailingAnchor),
+            secondaryNavigationController.view.topAnchor.constraint(equalTo: view.topAnchor),
+            view.trailingAnchor.constraint(equalTo: secondaryNavigationController.view.trailingAnchor),
+            view.bottomAnchor.constraint(equalTo: secondaryNavigationController.view.bottomAnchor)
+        ]
+        borderWidthConstraint = constraints[2]
+        primaryColumnWidthConstraint = constraints[7]
```
Comment (netbe) [inline WireUI/Sources/WireMainNavigationUI/Containers/DoubleColumnContainerViewController.swift]: what about creating first to the variable and then activate it later, I am afraid with the indexes this breaks easily:
```
borderWidthConstraint = borderView.widthAnchor.constraint(equalToConstant: borderWidth)

let constraints = [
...
borderWidthConstraint,
...
]
```

File: WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController.swift [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+import SwiftUI
+
+// swiftlint:disable opening_brace
+
+public final class MainSplitViewController<Sidebar, TabContainer>: UISplitViewController, MainSplitViewControllerProtocol where
+    Sidebar: MainSidebarProtocol,
+    TabContainer: MainTabBarControllerProtocol
+{
+    // swiftlint:enable opening_brace
```
Comment (netbe) [inline WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController.swift:27]: should it be enabled by default? needed?

File: WireUI/Sources/WireMainNavigationUI/Containers/MainTabBarController.swift [hunk 4]
```diff
@@ -88,12 +71,18 @@ public final class MainTabBarController<
...
 
-    private let conversationsNavigationController = UINavigationController()
-    private let archiveNavigationController = UINavigationController()
-    private let settingsNavigationController = UINavigationController()
+    private weak var _conversationList: ConversationList?
+    private weak var _folders: UIViewController?
+    private weak var _archive: Archive?
+    private weak var _settings: Settings?
+    private weak var _conversation: Conversation?
+    private weak var _settingsContent: UIViewController?
```
Comment (netbe) [inline WireUI/Sources/WireMainNavigationUI/Containers/MainTabBarController.swift]: question: why all these are weak?

File: WireUI/Sources/WireMainNavigationUI/Coordinator/AnyMainCoordinator.swift [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    private let _showConversation: @MainActor (_ conversation: ConversationModel, _ message: ConversationMessageModel?) async -> Void
+    private let _hideConversation: @MainActor () -> Void
+    private let _showSettingsContent: @MainActor (_ topLevelMenuItem: SettingsTopLevelMenuItem) -> Void
+    private let _hideSettingsContent: @MainActor () -> Void
+    private let _showSelfProfile: @MainActor () async -> Void
+    private let _showUserProfile: @MainActor (_ user: User) async -> Void
+    private let _showConnect: @MainActor () async -> Void
+    private let _showCreateGroupConversation: @MainActor () async -> Void
+    private let _presentViewController: @MainActor (_ viewController: UIViewController) async -> Void
+    private let _dismissPresentedViewController: @MainActor () async -> Void
```
Comment (netbe) [inline WireUI/Sources/WireMainNavigationUI/Coordinator/AnyMainCoordinator.swift:38]: is there a reason for prefixing _ to all these properties? how about making them directly public

File: WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController+typealiases.swift [hunk 6]
```diff
@@ -17,14 +17,17 @@
...
-    var settings: UIViewController?
-    var conversation: UIViewController?
-    var tabContainer: UIViewController!
+public extension MainSplitViewController {
+
+    typealias ConversationList = TabContainer.ConversationList
+    typealias Archive = TabContainer.Archive
+    typealias Settings = TabContainer.Settings
+
+    typealias Conversation = TabContainer.Conversation
```
Comment (netbe) [inline WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController+typealiases.swift]: maybe out of scope but this could be confusing to have it name Conversation as in my mind it relates to the data Conversation (ZMConversation)

File: WireUI/Sources/WireMainNavigationUI/Coordinator/MainCoordinatorDependencies.swift [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        ConnectBuilder.Dependencies == Self
+
+    associatedtype CreateGroupConversationBuilder: MainCoordinatorInjectingViewControllerBuilder where
+        CreateGroupConversationBuilder.Dependencies == Self
+
+    associatedtype SelfProfileBuilder: MainCoordinatorInjectingViewControllerBuilder where
+        SelfProfileBuilder.Dependencies == Self
+
+    associatedtype UserProfileBuilder: MainUserProfileBuilderProtocol where
+        UserProfileBuilder.User == User
```
Comment (netbe) [inline WireUI/Sources/WireMainNavigationUI/Coordinator/MainCoordinatorDependencies.swift]: not sure I understand why or how this works.

File: WireUI/Sources/WireMainNavigationUI/Preview/MainTabBarControllerPreview.swift [hunk 8]
```diff
@@ -36,7 +33,7 @@ private final class PlaceholderViewController: UIViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
 
-        view.backgroundColor = .yellow // ColorTheme.Backgrounds.surfaceVariant
+        view.backgroundColor = .yellow
```
Comment (netbe) [inline WireUI/Sources/WireMainNavigationUI/Preview/MainTabBarControllerPreview.swift:36]: is it supposed to be yellow?

File: WireUI/Sources/WireSidebarUI/Views/SidebarMenuItemView.swift [hunk 9]
```diff
@@ -50,16 +50,16 @@ struct SidebarMenuItemView: View {
     // MARK: -
 
     var body: some View {
-        let accentColor = wireAccentColorMapping.uiColor(for: wireAccentColor)
+        let accentColor = wireAccentColorMapping?.uiColor(for: wireAccentColor) ?? .systemGray
```
Comment (netbe) [inline WireUI/Sources/WireSidebarUI/Views/SidebarMenuItemView.swift]: why do we have this variable in body? can't we resolve it in init

File: WireUI/Sources/WireSidebarUI/Views/SidebarView.swift [hunk 10]
```diff
@@ -136,7 +136,35 @@ public struct SidebarView<AccountImageView>: View where AccountImageView: View {
...
 
-    private func menuItem(_ menuItem: SidebarMenuItem) -> some View {
+    private func nonselectableMenuItem(_ menuItem: SidebarNonselectableMenuItem) -> some View {
+        let text: Text
+        let icon: String
+        let isLink: Bool
+        let action: () -> Void
+        switch menuItem {
+        case .connect:
+            text = Text("sidebar.contacts.connect.title", bundle: .module)
```
Comment (netbe) [inline WireUI/Sources/WireSidebarUI/Views/SidebarView.swift:146]: sorry if I asked this before, but why we can't use swiftgen strings

File: WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift [hunk 11]
```diff
@@ -23,67 +23,72 @@ import XCTest
...
-        let sut = MainTabBarController<MockConversationListViewController>()
 
         // When
-        sut.conversations = (conversationList, conversation)
+        sut.conversationList = conversationList
 
         // Then
-        let navigationController = try XCTUnwrap(sut.viewControllers?[1] as? UINavigationController)
-        XCTAssertEqual(navigationController.viewControllers, [conversationList, conversation])
+        let navigationController = try XCTUnwrap(sut.viewControllers?[1] as? UINavigationController) // TODO: [WPB-6647] fix index for navigation overhaul
```
Comment (netbe) [inline WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift:53]: still valid TODO?

File: WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift [hunk 12]
```diff
@@ -23,67 +23,72 @@ import XCTest
...
 
         // When
-        navigationController.popViewController(animated: false)
+        sut.conversationList = nil
 
         // Then
-        let (conversationList, conversation) = try XCTUnwrap(sut.conversations)
-        XCTAssertNotNil(conversationList)
-        XCTAssertNil(conversation)
+        await Task.yield()
```
Comment (netbe) [inline WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift:71]: I am not sure to understand what we're doing here

File: WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift [hunk 13]
```diff
@@ -23,67 +23,72 @@ import XCTest
...
         // Given
         let archive = UIViewController()
-        let sut = MainTabBarController<MockConversationListViewController>()
 
         // When
         sut.archive = archive
 
         // Then
-        let navigationController = try XCTUnwrap(sut.viewControllers?[3] as? UINavigationController)
+        let navigationController = try XCTUnwrap(sut.viewControllers?[3] as? UINavigationController) // TODO: [WPB-6647] fix index for navigation overhaul
```
Comment (netbe) [inline WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift:84]: is this todo tackled in a later PR?

File: WireUI/Tests/WireMainNavigationUITests/Coordinator/AnyMainCoordinatorTests.swift [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        mockMainCoordinator = .init()
+        sut = .init(mainCoordinator: mockMainCoordinator)
+    }
+
+    override func tearDown() {
+        sut = nil
+        mockMainCoordinator = nil
+    }
+
+    @MainActor
```
Comment (netbe) [inline WireUI/Tests/WireMainNavigationUITests/Coordinator/AnyMainCoordinatorTests.swift]: at that point, is it worth making AnyMainCoordinatorTests a mainActor?

File: WireUI/Tests/WireMainNavigationUITests/Coordinator/MainCoordinatorTests.swift [hunk 15]
```diff
@@ -87,24 +93,24 @@ final class MainCoordinatorTests: XCTestCase {
...
 
         // Then
         XCTAssertNil(splitViewController.conversationList)
-        XCTAssertNotNil(tabBarController.conversations?.conversationList)
+        XCTAssertNotNil(tabBarController.conversationList)
         XCTAssertNotNil(splitViewController.archive)
         XCTAssertNil(tabBarController.archive)
     }
 
     // TODO: [WPB-10903] add many more tests, e.g.
```
Comment (netbe) [inline WireUI/Tests/WireMainNavigationUITests/Coordinator/MainCoordinatorTests.swift]: is this done in another PR?

File: WireUI/Tests/WireMainNavigationUITests/Mocks/MockMainCoordinator.swift [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+import WireMainNavigationUI
+
+final class MockMainCoordinatorProtocol: MainCoordinatorProtocol {
```
Comment (netbe) [inline WireUI/Tests/WireMainNavigationUITests/Mocks/MockMainCoordinator.swift:24]: this doesn't use sourcery, is there a reason why? if yes could you add it in comment on the class

File: WireUI/Tests/WireMainNavigationUITests/Mocks/MockMainCoordinatorDependencies.swift [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+    // MainCoordinatorProtocolDependencies
+
+    typealias ConversationFilter = MainConversationFilter
+    typealias ConversationModel = PreviewConversationModel
+    typealias ConversationMessageModel = Void
+    typealias SettingsTopLevelMenuItem = MainSettingsTopLevelMenuItem
+    typealias User = MockUser
+
+    // MainCoordinatorDependencies
```
Comment (netbe) [inline WireUI/Tests/WireMainNavigationUITests/Mocks/MockMainCoordinatorDependencies.swift]:
```suggestion
    // MARK: - MainCoordinatorDependencies
```

**Review by netbe — COMMENTED** [review]
I added some comments, but I feel a tour of the changes to explain it would be great going further

File: WireUI/Sources/WireDesign/Appearance/SidebarViewDesign.swift [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+    public let backgroundColor = ColorTheme.Backgrounds.background
+    public let accountInfoViewDisplayNameColor = ColorTheme.Backgrounds.onSurface
+    public let accountInfoViewUsernameColor = ColorTheme.Base.secondaryText
+    public let menuItemTitleForegroundColor = ColorTheme.Backgrounds.onBackground
+    public let menuItemLinkIconForegroundColor = ColorTheme.Base.secondaryText
+    public let menuItemIsSelectedTitleForegroundColor = ColorTheme.Base.onPrimary
+
+    public init() {}
+}
```
Comment (agisilaos) [inline WireUI/Sources/WireDesign/Appearance/SidebarViewDesign.swift:31]: I really like this pattern on how we define colors for a component!

File: WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController.swift (hunk 3 above)
Comment (agisilaos) [reply WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController.swift:27]: I think it should be enabled by default.

File: WireFoundation/Sources/WireTesting/SDKExtensions/UIKit/UISplitViewControllerDisplayMode+CustomDebugStringConvertible.swift (hunk 1 above)
Comment (caldrian) [reply WireFoundation/Sources/WireTesting/SDKExtensions/UIKit/UISplitViewControllerDisplayMode+CustomDebugStringConvertible.swift]: Unfortunately not:
```
let displayMode = UISplitViewController.DisplayMode.oneOverSecondary
print(String(reflecting: displayMode)) // "__C.UISplitViewControllerDisplayMode(rawValue: 3)"
print(String(describing: displayMode)) // "UISplitViewControllerDisplayMode(rawValue: 3)"
```

File: WireUI/Sources/WireMainNavigationUI/Containers/DoubleColumnContainerViewController.swift (hunk 2 above)
Comment (caldrian) [reply WireUI/Sources/WireMainNavigationUI/Containers/DoubleColumnContainerViewController.swift]: fixed: 5d54e4ae94a0067d99b0ae8c38d84fc3278e3702

File: WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController.swift (hunk 3 above)
Comment (caldrian) [reply WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController.swift:27]: This fixes a conflicting behavior between SwiftFormat and SwiftLint.

File: WireUI/Sources/WireDesign/Appearance/SidebarViewDesign.swift (hunk 18 above)
Comment (caldrian) [reply WireUI/Sources/WireDesign/Appearance/SidebarViewDesign.swift:31]: To be honest, I'm not sure yet. :D

File: WireUI/Tests/WireMainNavigationUITests/Mocks/MockMainCoordinator.swift (hunk 16 above)
Comment (caldrian) [reply WireUI/Tests/WireMainNavigationUITests/Mocks/MockMainCoordinator.swift:24]: I tried but it led to compile errors.
We have to fix our stencils. But even then I'm not sure if protocols with associated types work.

File: WireUI/Tests/WireMainNavigationUITests/Coordinator/MainCoordinatorTests.swift (hunk 15 above)
Comment (caldrian) [reply WireUI/Tests/WireMainNavigationUITests/Coordinator/MainCoordinatorTests.swift]: I had created a separate ticket to think about it again. Maybe it's not needed.

File: WireUI/Tests/WireMainNavigationUITests/Coordinator/AnyMainCoordinatorTests.swift (hunk 14 above)
Comment (caldrian) [reply WireUI/Tests/WireMainNavigationUITests/Coordinator/AnyMainCoordinatorTests.swift]: fixed: c4f678e363e2de2ff75c35d47bd9aa1da2ef8a02

File: WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift (hunk 13 above)
Comment (caldrian) [reply WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift:84]: Yes, in the main PR for the epic branch: https://github.com/wireapp/wire-ios/pull/1361
On develop we still show four tabs.

File: WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift (hunk 12 above)
Comment (caldrian) [reply WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift:71]: I wanted to ensure that assigning `nil` to any of the properties actually releases the previous reference (since internally the reference needs to be removed from the view controller hierarchy).

File: WireUI/Sources/WireMainNavigationUI/Preview/MainTabBarControllerPreview.swift (hunk 8 above)
Comment (caldrian) [reply WireUI/Sources/WireMainNavigationUI/Preview/MainTabBarControllerPreview.swift:36]: No, this is just preview code (see the file name).
Sometimes I use colors for better visibility in previews and/or snapshot tests.
Example:
Screenshot 2024-10-16 at 12 48 47 PM

File: WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift (hunk 11 above)
Comment (caldrian) [reply WireUI/Tests/WireMainNavigationUITests/Containers/MainTabBarControllerTests.swift:53]: Yes, since this PR merges only parts of the navigation overhaul to develop.
The remaining changes are part of the epic pr https://github.com/wireapp/wire-ios/pull/1361 which should remove all the TODOs with the ticket number `WPB-6647`.

File: WireUI/Sources/WireSidebarUI/Views/SidebarView.swift (hunk 10 above)
Comment (caldrian) [reply WireUI/Sources/WireSidebarUI/Views/SidebarView.swift:146]: I had issues setting up SwiftGen for our packages. I created a ticket to fix it later: https://wearezeta.atlassian.net/browse/WPB-11610

File: WireUI/Sources/WireSidebarUI/Views/SidebarMenuItemView.swift (hunk 9 above)
Comment (caldrian) [reply WireUI/Sources/WireSidebarUI/Views/SidebarMenuItemView.swift]: fixed: f41bedb14c990c99e1f03144e449f010c2a2def7

File: WireUI/Sources/WireMainNavigationUI/Coordinator/MainCoordinatorDependencies.swift (hunk 7 above)
Comment (caldrian) [reply WireUI/Sources/WireMainNavigationUI/Coordinator/MainCoordinatorDependencies.swift]: Discussed in a call.

File: WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController+typealiases.swift (hunk 6 above)
Comment (caldrian) [reply WireUI/Sources/WireMainNavigationUI/Containers/MainSplitViewController+typealiases.swift]: I added a "UI" suffix to all these types.

File: WireUI/Sources/WireMainNavigationUI/Coordinator/AnyMainCoordinator.swift (hunk 5 above)
Comment (caldrian) [reply WireUI/Sources/WireMainNavigationUI/Coordinator/AnyMainCoordinator.swift:38]: Making them public doesn't satisfy the protocol requirements.
I could have chosen other names so that they don't collide with the method names, however, I think the prefix-underscore is quite common in Swift and ObjC. (e.g. in property wrappers)

File: WireUI/Sources/WireMainNavigationUI/Containers/MainTabBarController.swift (hunk 4 above)
Comment (caldrian) [reply WireUI/Sources/WireMainNavigationUI/Containers/MainTabBarController.swift]: For `conversationUI` and `settingsContentUI` it makes the most sense, since the user can tap or swipe back and the UINavigationController dismisses the view controller. If we kept a strong reference here, we'd prevent the VC from being deallocated. 
For the other properties it probably makes no difference.

**Review by agisilaos — APPROVED** [review]

File: WireFoundation/Sources/WireTesting/SDKExtensions/UIKit/UISplitViewControllerDisplayMode+CustomDebugStringConvertible.swift (hunk 1 above)
Comment (netbe) [reply WireFoundation/Sources/WireTesting/SDKExtensions/UIKit/UISplitViewControllerDisplayMode+CustomDebugStringConvertible.swift]: ok

**Review by netbe — APPROVED** [review]

File: WireUI/Tests/WireMainNavigationUITests/Mocks/MockMainCoordinator.swift (hunk 16 above)
Comment (netbe) [inline WireUI/Tests/WireMainNavigationUITests/Mocks/MockMainCoordinator.swift:24]:
```suggestion
/// Note: This mock is manual because it let to compile errors (generics?)
final class MockMainCoordinatorProtocol: MainCoordinatorProtocol {
```
